File: mergeSort.c
#include<stdio.h>
#include<stdlib.h>

void merge(int *p,int left,int mid,int right)
{
    int ini;
    int lSize = mid-left+1;
    int rSize = right-mid;
    int *l = (int*)malloc(lSize*sizeof(int));
    int *r = (int*)malloc(rSize*sizeof(int));
    for (ini=0;ini<lSize;ini++)
    {
        *(l+ini)= *(p+left+ini);
    }
    for (ini=0;ini<rSize;ini++)
    {
        *(r+ini)= *(p+mid+1+ini);
    }

    int i=0,j=0,k=left;
    // printf("merging");
    while(i<lSize && j<rSize)
    {
        if(*(l+i)>=*(r+j))
        {
            *(p+k) = *(r+j); //if ascending 
            //*(p+k) = *(l+i); if descending
            j++;
        }
        else 
        {
            *(p+k) = *(l+i); //if ascending
            //*(p+k) = *(r+j); if descending 
            i++;
        }
        k++;
    }
    while(i<lSize)
    {
        *(p+k) = *(l+i);
        i++;
        k++;
    }
    while(j<rSize)
    {
        *(p+k) = *(r+j);
        j++;
        k++;
    }
}


void mergeSort(int *p, int lIndex,int rIndex)
{
    int l,r,m;
    l = lIndex;
    r = rIndex;
    if(l<r)
    {
        m = l+(r-l)/2;
        mergeSort(p,l,m);
        mergeSort(p,m+1,r);
        merge(p,l,m,r);
    }
}


int main()
{
    int x,n,i,*p;
    printf("Enter the number of elements in the array: ");
    scanf("%d",&n);
    p=(int*)malloc(n*sizeof(int));
    printf("Enter the elements of the array: ");
    for(i=0;i<n;i++)
    {
        scanf("%d",&p[i]);
    }
    mergeSort(p,0,n-1);
    printf("sorted array is \n");
    for(i=0;i<n;i++)
    {
        printf("%d     ",p[i]);
    }

    return 0;
}

File: mergesort.py
def merge_sort(number_list,order=0):
    if len(number_list)>1:
        mid = len(number_list) // 2
        left_arr = number_list[:mid]
        right_arr = number_list[mid:]
        merge_sort(left_arr,order)
        merge_sort(right_arr,order)

        i=j=k=0
        while i<len(left_arr) and j<len(right_arr):
            if left_arr[i]<=right_arr[j]:
                if order==0:
                    number_list[k]=left_arr[i]
                    i+=1
                else:
                    number_list[k]=right_arr[j]
                    j+=1
            else:
                if order==0:
                    number_list[k]=right_arr[j]
                    j+=1
                else:
                    number_list[k]=left_arr[i]
                    i+=1
            k+=1
        while i<len(left_arr):
            number_list[k]=left_arr[i]
            i+=1
            k+=1
        while j<len(right_arr):
            number_list[k]=right_arr[j]
            j+=1
            k+=1

number_list = list(map(int,input("Enter a list of numbers: ").strip().split()))
merge_sort(number_list)
print("after descending sort ")
print (number_list) 
print("after ascending sort ")
merge_sort(number_list,order=1)
print (number_list) 